arch/arm/src/stm32wl5/stm32wl5_spi.c

```diff
@@ -0,0 +1,1858 @@
+   *  frames by a single 16-bit read of SPIx_DR as a response to this single
+   *  RXNE event. The RxFIFO threshold setting and the following read access
+   *  must be always kept aligned at the receiver side, as data can be lost
+   *  if it is not in line."
+   */
+
+  if (priv->nbits < 9)
+    {
+      return (uint16_t)spi_getreg8(priv, STM32WL5_SPI_DR_OFFSET);
+    }
+  else
+    {
+      return spi_getreg(priv, STM32WL5_SPI_DR_OFFSET);
+    }
+}
+
+/****************************************************************************
+ * Name: spi_writeword
+ *
+ * Description:
+ *   Write one word or byte to SPI. If the frame size is 8 bit or lower
+ *   a byte is written. In other case a word is written.
+ *
+ * Input Parameters:
+ *   priv - Device-specific state data
+ *   word - word to send
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static inline void spi_writeword(struct stm32wl5_spidev_s *priv,
+                                 uint16_t word)
+{
+  /* Wait until the transmit buffer is empty */
+
+  while ((spi_getreg(priv, STM32WL5_SPI_SR_OFFSET) & SPI_SR_TXE) == 0)
+    {
+    }
+
+  /* Then send the word */
+
+  if (priv->nbits < 9)
+    {
+      spi_putreg8(priv, STM32WL5_SPI_DR_OFFSET, (uint8_t)word);
+    }
+  else
+    {
+      spi_putreg(priv, STM32WL5_SPI_DR_OFFSET, word);
+    }
+}
+
+/****************************************************************************
+ * Name: spi_dmarxwait
+ *
+ * Description:
+ *   Wait for DMA to complete.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_STM32WL5_SPI_DMA
+static int spi_dmarxwait(struct stm32wl5_spidev_s *priv)
+{
+  int ret;
+
+  /* Take the semaphore (perhaps waiting).  If the result is zero, then the
+   * DMA must not really have completed???
+   */
+
+  do
+    {
+      ret = nxsem_wait_uninterruptible(&priv->rxsem);
+
+      /* The only expected error is ECANCELED which would occur if the
+       * calling thread were canceled.
+       */
+
+      DEBUGASSERT(ret == OK || ret == -ECANCELED);
+    }
+  while (priv->rxresult == 0 && ret == OK);
+
+  return ret;
+}
+#endif
+
+/****************************************************************************
+ * Name: spi_dmatxwait
+ *
+ * Description:
+ *   Wait for DMA to complete.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_STM32WL5_SPI_DMA
+static int spi_dmatxwait(struct stm32wl5_spidev_s *priv)
+{
+  int ret;
+
+  /* Take the semaphore (perhaps waiting).  If the result is zero, then the
+   * DMA must not really have completed???
+   */
+
+  do
+    {
+      ret = nxsem_wait_uninterruptible(&priv->txsem);
+
+      /* The only expected error is ECANCELED which would occur if the
+       * calling thread were canceled.
+       */
+
+      DEBUGASSERT(ret == OK || ret == -ECANCELED);
+    }
+  while (priv->txresult == 0 && ret == OK);
```

Comment:
```suggestion
  while (priv->txresult == 0 && ret != OK);
```

Reply:
Let me re-examine this part of code again, but I think the code expects to wait for DMA transaction to complete, so even if task is interrupted due to cancel request we it should resume waiting

Reply:
Code before change was a deadloop in case if wait returns `OK`. Now I think it should be fine. If you do not like `ret != OK` we can change it to `ret == -ECANCELED`

Reply:
But if the thread is cancelled, then it does not signal to semaphore and the  `nxsem_wait_uninterruptible` will wait forever?

Reply:
I see the semaphore gets signal from `spi_dmarxcallback`, which is caused by interrupt, so should work.  But I do not have an idea what would be a reliable code...

Reply:
@pkarashchenko 
There is a [Cancellation Points](https://cwiki.apache.org/confluence/display/NUTTX/Cancellation+Points) notes where the case is described. So the condition `while (priv->txresult == 0 && ret == OK);` looks correct for me.